src/emulator.rs

```diff
@@ -79,11 +79,35 @@ impl Cpu<u64> {
     // #### Control
 
     // `beq rs1,rs2,imm`: `if (rs1 == rs2) { pc = pc + imm } else { pc = pc + 4 }` with `-2^12 <= imm < 2^12` and `imm % 2 == 0`
-    // TODO
+    pub fn beq(&mut self, rs1: usize, rs2: usize, imm: i64) {
+        if self.regs[rs1] == self.regs[rs2] {
+            assert!(imm < 2 << 12 && imm >= -(2 << 12));
+            assert_eq!(imm % 2, 0);
+            assert_eq!(imm % 4, 0, "instruction-address-misaligned");
+            let pc = self.pc as i128;
+            self.pc = (pc + imm as i128) as u64;
+        } else {
+            self.pc = self.pc + 4;
+        }
+    }
     // `jal rd,imm`: `rd = pc + 4; pc = pc + imm` with `-2^20 <= imm < 2^20` and `imm % 2 == 0`
-    // TODO
+    pub fn jal(&mut self, rd: usize, imm: i64) {
+        assert!(imm < 2 << 12 && imm >= -(2 << 12));
```

Comment:
From the RISC-V spec, the offset from `imm` is 20 bits and has a longer range than in BEQ:
> Jumps can therefore target a ±1 MiB range

```suggestion
        assert!(imm < 2 << 20 && imm >= -(2 << 20));
```